// @auth0/auth0-fastify-api/dist/index.js
// src/index.ts
import fp from "fastify-plugin";
import { ApiClient } from "@auth0/auth0-api-js";
function validateScopes(token, requiredScopes) {
  const scopes = Array.isArray(requiredScopes) ? requiredScopes : [requiredScopes];
  let tokenScopes = [];
  if (token.scope) {
    tokenScopes = typeof token.scope === "string" ? token.scope.split(" ") : token.scope;
  }
  return scopes.every((required) => tokenScopes.includes(required));
}
async function auth0FastifyApi(fastify, options) {
  if (!options.audience) {
    throw new Error("In order to use the Auth0 Api plugin, you must provide an audience.");
  }
  const apiClient = new ApiClient({
    domain: options.domain,
    audience: options.audience,
    clientId: options.clientId,
    clientSecret: options.clientSecret,
    clientAssertionSigningKey: options.clientAssertionSigningKey,
    clientAssertionSigningAlg: options.clientAssertionSigningAlg,
    customFetch: options.customFetch
  });
  const replyWithError = (reply, statusCode, error, errorDescription) => {
    return reply.code(statusCode).header(
      "WWW-Authenticate",
      `Bearer error="${error.replaceAll('"', '\\"')}", error_description="${errorDescription.replaceAll('"', '\\"')}"`
    ).send({
      error,
      error_description: errorDescription
    });
  };
  fastify.decorate("requireAuth", function(opts = {}) {
    return async function(request, reply) {
      const accessToken = getToken(request);
      if (!accessToken) {
        return replyWithError(reply, 400, "invalid_request", "No Authorization provided");
      }
      try {
        const token = await apiClient.verifyAccessToken({ accessToken });
        if (opts.scopes && !validateScopes(token, opts.scopes)) {
          return replyWithError(reply, 403, "insufficient_scope", "Insufficient scopes");
        }
        request["user"] = token;
      } catch (error) {
        if (error.code === "verify_access_token_error") {
          return replyWithError(reply, 401, "invalid_token", error.message);
        }
        return replyWithError(reply, 401, "invalid_token", "Invalid token");
      }
    };
  });
  fastify.decorateRequest("getToken", function() {
    return getToken(this);
  });
  fastify.decorate("auth0Client", apiClient);
}
var index_default = fp(auth0FastifyApi);
function getToken(request) {
  const parts = request.headers.authorization?.split(" ");
  return parts?.length === 2 && parts[0]?.toLowerCase() === "bearer" ? parts[1] : void 0;
}
export {
  index_default as default
};
