// @auth0/auth0-fastify-api/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  default: () => index_default
});
module.exports = __toCommonJS(index_exports);
var import_fastify_plugin = __toESM(require("fastify-plugin"), 1);
var import_auth0_api_js = require("@auth0/auth0-api-js");
function validateScopes(token, requiredScopes) {
  const scopes = Array.isArray(requiredScopes) ? requiredScopes : [requiredScopes];
  let tokenScopes = [];
  if (token.scope) {
    tokenScopes = typeof token.scope === "string" ? token.scope.split(" ") : token.scope;
  }
  return scopes.every((required) => tokenScopes.includes(required));
}
async function auth0FastifyApi(fastify, options) {
  if (!options.audience) {
    throw new Error("In order to use the Auth0 Api plugin, you must provide an audience.");
  }
  const apiClient = new import_auth0_api_js.ApiClient({
    domain: options.domain,
    audience: options.audience,
    clientId: options.clientId,
    clientSecret: options.clientSecret,
    clientAssertionSigningKey: options.clientAssertionSigningKey,
    clientAssertionSigningAlg: options.clientAssertionSigningAlg,
    customFetch: options.customFetch
  });
  const replyWithError = (reply, statusCode, error, errorDescription) => {
    return reply.code(statusCode).header(
      "WWW-Authenticate",
      `Bearer error="${error.replaceAll('"', '\\"')}", error_description="${errorDescription.replaceAll('"', '\\"')}"`
    ).send({
      error,
      error_description: errorDescription
    });
  };
  fastify.decorate("requireAuth", function(opts = {}) {
    return async function(request, reply) {
      const accessToken = getToken(request);
      if (!accessToken) {
        return replyWithError(reply, 400, "invalid_request", "No Authorization provided");
      }
      try {
        const token = await apiClient.verifyAccessToken({ accessToken });
        if (opts.scopes && !validateScopes(token, opts.scopes)) {
          return replyWithError(reply, 403, "insufficient_scope", "Insufficient scopes");
        }
        request["user"] = token;
      } catch (error) {
        if (error.code === "verify_access_token_error") {
          return replyWithError(reply, 401, "invalid_token", error.message);
        }
        return replyWithError(reply, 401, "invalid_token", "Invalid token");
      }
    };
  });
  fastify.decorateRequest("getToken", function() {
    return getToken(this);
  });
  fastify.decorate("auth0Client", apiClient);
}
var index_default = (0, import_fastify_plugin.default)(auth0FastifyApi);
function getToken(request) {
  const parts = request.headers.authorization?.split(" ");
  return parts?.length === 2 && parts[0]?.toLowerCase() === "bearer" ? parts[1] : void 0;
}
